api/models/datastore.go

```diff
@@ -5,21 +5,66 @@ import (
 	"errors"
 )
 
+//TODO path name validation: implementations currently assume/require/rely on:
+// - leading slash
+// - overall legal/scrubbed path (no relative paths, consecutive slashes, etc)
+// - trailing slash allowed and distinct from none
 type Datastore interface {
-	// GetApp returns the app called appName or nil if it doesn't exist
+
+	// GetApp gets an App by name.
+	// Returns ErrDatastoreEmptyAppName for empty appName.
+	// Returns ErrAppsNotFound if no app is found.
 	GetApp(ctx context.Context, appName string) (*App, error)
+
+	// GetApps gets a slice of Apps, optionally filtered by name.
+	// Missing filter or empty name will match all Apps.
 	GetApps(ctx context.Context, filter *AppFilter) ([]*App, error)
+
+	// InsertApp inserts an App. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsAlreadyExists if an App by the same name already exists.
 	InsertApp(ctx context.Context, app *App) (*App, error)
+
+	// UpdateApp updates an App's Config. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsNotFound if an App is not found.
+	//TODO are app.Config's fields added to the existing field or do they replace them?
 	UpdateApp(ctx context.Context, app *App) (*App, error)
```

Comment:
Patching ```map[string]string``` types is straightforward, since empty string values can signal deletes.  However, ```Route.Header``` is ```map[string][]string```, so assuming we patch the values by concatenating the arrays (or set union), what would be the protocol for deleting keys or individual value array elements?

I suppose an empty array value could signal 'delete key', but it would be unfortunate to have to nuke a key and reinsert all but one array element in order to remove a single element.

Am I overlooking something simple?